src/charm.py

```diff
@@ -2,58 +2,77 @@
 # Copyright 2021 Chris
 # See LICENSE file for licensing details.
 
-"""Charm the service.
-
-Refer to the following post for a quick-start guide that will help you
-develop a new k8s charm using the Operator Framework:
-
-    https://discourse.charmhub.io/t/4208
-"""
-
 import logging
+import subprocess
 
 from ops.charm import CharmBase
 from ops.framework import StoredState
 from ops.main import main
-from ops.model import ActiveStatus
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus
+from runner import Runner
 
 logger = logging.getLogger(__name__)
 
 
 class GithubRunnerOperator(CharmBase):
-    """Charm the service."""
-
     _stored = StoredState()
 
     def __init__(self, *args):
         super().__init__(*args)
         self.framework.observe(self.on.install, self._on_install)
-        self.framework.observe(self.on.config_changed, self._on_config_changed)
-        self.framework.observe(self.on.fortune_action, self._on_fortune_action)
-        self._stored.set_default(things=[])
+        # self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.register_action, self._on_register_action)
+        self._stored.set_default(registered=False)
+        self._runner = Runner()
 
     def _on_install(self, _):
-        self.unit.status = ActiveStatus("Active and running")
-
-    def _on_config_changed(self, _):
-        # Note: you need to uncomment the example in the config.yaml file for this to work (ensure
-        # to not just leave the example, but adapt to your configuration needs)
-        current = self.config["thing"]
-        if current not in self._stored.things:
-            logger.debug("found a new thing: %r", current)
-            self._stored.things.append(current)
-
-    def _on_fortune_action(self, event):
-        # Note: you need to uncomment the example in the actions.yaml file for this to work (ensure
-        # to not just leave the example, but adapt to your needs for actions commands)
-        fail = event.params["fail"]
-        if fail:
-            event.fail(fail)
-        else:
-            event.set_results(
-                {"fortune": "A bug in the code is worth two in the documentation."}
-            )
+        self.unit.status = MaintenanceStatus("Installing runner")
+        self._runner.download()
+        self._runner.setup_env()
+        self.unit.status = BlockedStatus("Waiting for registration")
+
+    def _on_register_action(self, event):
+        if self._stored.registered:
+            pass  # TODO raise this an an error
```

Comment:
Why not do this now?

Reply:
Yup I'll add a log and return the status.